### ISSUE TITLE
Qubes Debian 9 VM crash during pupy installation.

### ISSUE THREAD

**Issue by Strazzom** [n1nj4sec/pupy#645]
This is a new issue, given that this is a different problem than originally documented.

These has already been extensive conversation about this in #639, as well as documented steps to reproduce on [this](https://github.com/n1nj4sec/pupy/wiki/Qubes-specific-Issues-Bug-Testing-Report---August-21,-2018) wiki page.

This is NOT the same issue where the linux source build fails with an error. This is an outright crash during the step `[+] Install python packages` on Qubes 3.2, with a Debian 9 PV VM.

**Comment by AlphaMufasaOmega** [comment]
I successfully executed: https://github.com/n1nj4sec/pupy/wiki/Installation-(Debian-9) in Qubes 4.0 by: enlarging my memory size in qubes VM 
installing `jre 10.*.*`(most updated version)
and maybe also `jarsigner` as a package individually 
executing the command `qvm-prefs -s <vm-name> kernelopts "nopat vsyscall=emulate"` ine the Dom0 terminal in order to configure the VM properly and then going through the steps detailed above or below
` https://github.com/n1nj4sec/pupy/wiki/Installation-(Debian-9)`

**Comment by AlphaMufasaOmega** [comment]
thanks guys you are the best! this is a great addition of capability to pupy! now i have to test the generation of those .apk files...

**Comment by Strazzom** [comment]
@AlphaMufasaOmega Can you please paste the exact commands you used to install `jre 10.*.*` and `jarsigner`?

**Comment by Strazzom** [comment]
On a completely unrelated note: anyone reading this who uses Qubes 3.2, can you please try to build the docker container by following the Debian 9 installation guide on the wiki [here](https://github.com/n1nj4sec/pupy/wiki/Installation-(Debian-9)), after running: `qvm-prefs -s <vm-name> kernelopts "nopat vsyscall=emulate"`?

I am at a complete loss as to why this VM keeps crashing. If anyone besides me can try to reproduce this issue, it would be immensely helpful.

**Comment by alxchk** [comment]
> I am at a complete loss as to why this VM keeps crashing. 

Try to debug this. Don't use build-docker.sh, manually repeate commands. Try to add strace during executing things at container. Read the logs (dmesg, syslog etc). Save strace log and so on

**Comment by AlphaMufasaOmega** [comment]
First and foremost on a **fully updated Qubes 4.0 Debian-9 TEMPLATE** do the following:

1.execute the command:
 `qvm-prefs -s <vm-name> kernelopts "nopat vsyscall=emulate"` in the Dom0 terminal in order to configure the VM (put name of VM in <vm-name)

2.Change the ram in the VM settings and make sure that their is  plenty of ram **AND ALSO** disk space for system storage **AND** private storage, you will get errors at step#5 and also earlier if you don't.

3. Install the JDK 10.*.*(most updated version) by following the directions below:
This is the link to the directions(it has screenshots):`https://linuxhint.com/install-openjdk-10-debian-9-stretch/`
Below is the directions that I followed for installing the JDK:

```
In short, Java Development Kit (JDK) is used to compile Java source code into Java byte code. Then Java Runtime Environment (JRE) translates the Java byte code into machine language and runs Java programs. If you install JDK, you don’t have to install JRE separately. But if you wish to only run Java programs, not to develop them, then you may install it separately.

There are 2 versions of JDK. Oracle JDK and OpenJDK. All the components of OpenJDK are open source. The same does not go with Oracle JDK. Some components of Oracle JDK are proprietary. This is why some people prefer OpenJDK over Oracle JDK.

OpenJDK 10 recently just came out. So in this article I am going to show you how to install it in Debian 9 Stretch. Let’s get started.
Downloading OpenJDK 10

In this section, I will show you how to download OpenJDK 10 from the official website of OpenJDK.

First go to the official website of OpenJDK 10 at http://jdk.java.net/10/ and you should see the following page as shown in the screenshot below.

Scroll down a little bit to the Downloads section. Click on the link as marked in the screenshot below to download the 64-bit Linux version of OpenJDK 10 tar.gz archive.

Click on Save File and then click on OK.

Your download should start as shown in the screenshot below.

Installing OpenJDK 10

Once your download is complete, you should find it in the default Downloads directory of your web browsers. In my case, it’s the Downloads/ directory in my user’s HOME directory.

First, navigate to the Downloads/ directory with the following command:
$ cd ~/Downloads

Now if you list the directory contents with ls -lh command, you should find the OpenJDK 10 tar archive as shown in the marked section of the screenshot below.

Now you have to extract the tar archive. I am going to extract it in my /opt directory. You may wish to choose another directory. You’re free to do so.
$ sudo tar xvzf openjdk-10_linux-x64_bin.tar.gz -C INSTALL_DIR

The tar archive should be extracted.

Now if you list your INSTALL_DIR, in my case it’s the /opt directory, you should see jdk-10/ directory as marked in the screenshot below.

Now you have to add it to the PATH of your Debian 9 operating system. That way you don’t have to type in the whole path to your OpenJDK 10 binary files in order to compile or run Java programs.

Run the following command to create a new file jdk10.sh in /etc/profile.d directory:
$ sudo nano /etc/profile.d/jdk10.sh

You should see the following window.

Now add the following lines to the file:
export JAVA_HOME="NSTALL_DIR/jdk-10"
export PATH="$PATH:${JAVA_HOME}/bin"

The INSTALL_DIR in my case it /opt

Now press <Ctrl> + x, then press y and then press <Enter> to save the file.

Once the file is saved, reboot your computer with the following command:
$ sudo reboot

Once your computer starts, run the following command to test whether the JAVA_HOME environment variable is available:
$ echo $JAVA_HOME

As you can see in the screenshot below, it’s set to the correct path.

You can also check the PATH environment variable with the following command:
$ echo $PATH

As you can see in the marked section of the screenshot below, the OpenJDK 10’s bin/ directory is added to the PATH.

Now run the following command to check whether OpenJDK is working or not:
$ javac -version

As you can see in the screenshot below, OpenJDK 10 is working correctly.

Compiling and Running a Java Program with OpenJDK 10

In this section, I will compile and run a simple Java program with OpenJDK 10 to show you that OpenJDK 10 is working correctly.

This is my simple hello world program saved in ~/Documents/codes/HelloWorld.java file.
public class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello World from LinuxHint!");
    }
}

Compiling HelloWorld.java with OpenJDK 10

First navigate to the directory where the file is saved with the following command:
$ cd ~/Documents/codes

Now to compile HelloWorld.java Java source file, run the following command:
$ javac HelloWorld.java

Running HelloWorld with OpenJDK 10

Once the HelloWorld.java file is compiled, a HelloWorld.class file is generated. This is the Java byte code file that you run with the java command as follows:
$ java HelloWorld

NOTE: If the Java byte code filename is HelloWorld.class, you run java HelloWorld. You must not include .class extension. Otherwise it won’t work.

As you can see, the Java program ran successfully and displayed the correct output.

So that’s how you install OpenJDK 10 on Debian 9 Stretch. Thank you for reading this article. 
```

4. After this Follow the instructions provided by the pupy team


```

The original instructions that these are based on can be found here.

The following commands should be executed as the root user:

apt-get update && apt-get install curl -y
curl -fsSL https://get.docker.com > docker_installer.sh
chmod +x ./docker_installer.sh && ./docker_installer.sh

Before executing the file, please verify that the content has not been tampered with in transit, as there is no automatic integrity checking.

For the following command, the actual username should be substituted for <username>.

usermod -aG docker <username>

Pupysh system-wide dependencies

The following commands should be executed as the root user:

apt-get install git libssl1.0-dev libffi-dev python-dev \
	python-pip tcpdump python-virtualenv

Pupy setup

The following commands should be executed as a non-root user:

The script pupy/create-workspace.py will create a Python virutalenv in the selected folder (pupyws in our example). The script also will create symlinks to pupysh at ~/.local/bin.

git clone --recursive https://github.com/alxchk/pupy
```

5.execute the command:
`cd pupy && sudo ./create-workspace.py pupyws`
(sudo is required due to the Qubes 4.0 environment;will get error if not enough ram and storage space in VM settings)

**Comment by Strazzom** [comment]
Thank you for the detailed walkthrough. I will attempt to replicate it.

Let's resume the conversation about the jarsigner issue over at #639.

**Comment by Strazzom** [comment]
I am going to close this, as it is no longer relevant with the new compose support and template downloading method. This will still probably remain an issue on Qubes, but only for people who want to compile the payload templates themselves, which I removed from the compose install for convenience. I may also remove it from the virtualenv install, and have it download the templates instead of compiling them.
